File: src/Tree/Tree.cs
#nullable enable
using System;
using System.Collections.Generic;
using System.Linq;

namespace Tree
{
	public sealed class Tree<T>
	{
		public TreeRootNode<T> Root { get; }

		public Tree(T rootValue) => Root = new TreeRootNode<T>(rootValue);

		public IEnumerable<(TreeNodeBase<T> Node, UInt32 Level)> TraverseDepthFirst()
		{
			yield return (Root, 0u);
			foreach (var (Node, Level) in Root.TraverseDescendantsDepthFirst())
				yield return (Node, Level + 1);
		}

		public IEnumerable<(TreeNodeBase<T> Node, UInt32 Level)> TraverseBreadthFirst()
		{
			yield return (Root, 0u);
			foreach (var (Node, Level) in Root.TraverseDescendantsBreadthFirst())
				yield return (Node, Level + 1);
		}
	}
}

File: src/typescript/Tree/Tree.ts
import { TreeRootNode, NodeLevel } from "./TreeNode";

export class Tree<T> {
	readonly root: TreeRootNode<T>;
	
	constructor(rootValue: T) {
		this.root = new TreeRootNode<T>(rootValue);
	}

	* traverseDepthFirst(): IterableIterator<NodeLevel<T>> {
		yield { node: this.root, level: 0 };
		for (let nodeLevel of this.root.traverseDescendantsDepthFirst())
			yield  { node: nodeLevel.node, level: nodeLevel.level + 1 };
	}

	* traverseBreadthFirst(): IterableIterator<NodeLevel<T>> {
		yield { node: this.root, level: 0 };
		for (let nodeLevel of this.root.traverseDescendantsBreadthFirst())
			yield  { node: nodeLevel.node, level: nodeLevel.level + 1 };
	}
}